sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,530 @@
+    private Channel channel;
+    private QueueingConsumer consumer;
+    private Instant currentTimestamp;
+    private final RabbitMQCheckpointMark checkpointMark;
+
+    UnboundedRabbitMqReader(RabbitMQSource source, RabbitMQCheckpointMark checkpointMark) {
+      this.source = source;
+      this.current = null;
+      this.checkpointMark = checkpointMark != null ? checkpointMark : new RabbitMQCheckpointMark();
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return checkpointMark.oldestTimestamp;
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return checkpointMark;
+    }
+
+    @Override
+    public RabbitMQSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public byte[] getCurrentRecordId() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      if (currentRecordId != null) {
+        return currentRecordId;
+      } else {
+        return "".getBytes(StandardCharsets.UTF_8);
+      }
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (currentTimestamp == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public RabbitMqMessage getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      try {
+        ConnectionFactory connectionFactory = createConnectionFactory(source.spec.uri());
+        connection = connectionFactory.newConnection();
+        channel = connection.createChannel();
+        if (channel == null) {
+          throw new IOException("No RabbitMQ channel available");
+        }
+
+        String queueName = source.spec.queue();
+        if (source.spec.queueDeclare()) {
+          // declare the queue (if not done by another application)
+          // channel.queueDeclare(queueName, durable, exclusive, autoDelete, arguments);
+          channel.queueDeclare(queueName, false, false, false, null);
+        }
+        if (source.spec.exchange() != null) {
+          channel.exchangeDeclare(source.spec.exchange(), source.spec.exchangeType());
+          if (queueName == null) {
+            queueName = channel.queueDeclare().getQueue();
+          }
+          channel.queueBind(queueName, source.spec.exchange(), source.spec.routingKey());
+        }
+        checkpointMark.channel = channel;
+        consumer = new QueueingConsumer(channel);
+        channel.txSelect();
+        // we consume message without autoAck (we want to do the ack ourselves)
+        channel.basicConsume(queueName, false, consumer);
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+      return advance();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      try {
+        QueueingConsumer.Delivery delivery = consumer.nextDelivery();
```

Comment:
Will this block forever if there's no next delivery? We should probably use a short timeout instead.

Reply:
Still relevant

Reply:
Yes, please set a CONST timeout. No need to expose the timeout to the user I think